service/src/main/java/uk/nhs/adaptors/oneoneone/cda/report/mapper/CarePlanMapper.java

```diff
@@ -0,0 +1,129 @@
+import static org.hl7.fhir.dstu3.model.IdType.newRandomUuid;
+import static uk.nhs.adaptors.oneoneone.cda.report.util.ReferenceUtil.ofTypes;
+
+@Component
+@AllArgsConstructor
+public class CarePlanMapper {
+    private final String SNOMED                   = "2.16.840.1.113883.2.1.3.2.4.15";
+    private final String INFORMATION_ADVICE_GIVEN = "1052951000000105";
+
+    private List<CarePlan> carePlansCreated = new ArrayList<>();
+
+    public List<Reference> mapCarePlan(POCDMT000002UK01ClinicalDocument1 clinicalDocument,
+                                       Encounter encounter) {
+
+        POCDMT000002UK01StructuredBody structuredBody = SectionUtil.getStructuredBody(clinicalDocument);
+
+        return Arrays.stream(structuredBody.getComponentArray())
+                .map(POCDMT000002UK01Component3::getSection)
+                .map(this::findCarePlanSections)
+                .flatMap(List::stream)
+                .map(section -> createCarePlanFromSection(section, encounter))
+                .map(cp -> create(cp))
+                .collect(Collectors.toUnmodifiableList());
+    }
+
+    public CarePlan createCarePlanFromSection(POCDMT000002UK01Section cpSection, Encounter encounter) {
+        CarePlan carePlan = new CarePlan();
+        carePlan.setIdElement(newRandomUuid());
+
+        if (cpSection.isSetLanguageCode()) {
+            carePlan.setLanguage(NodeUtil.getNodeValueString(cpSection.getLanguageCode()));
+        }
+        carePlan.setTitle(NodeUtil.getNodeValueString(cpSection.getTitle()));
+        String cpTextContent = NodeUtil.getNodeValueString(cpSection.getText().getContentArray(0));
+
+        Narrative narrative = new Narrative();
+        narrative.setDivAsString(cpTextContent);
+        carePlan.setText(narrative);
+        carePlan.setDescription(cpTextContent);
+        carePlan
+            .setIntent(PLAN)
+            .setSubject(encounter.getSubject())
+            .setSubjectTarget(encounter.getSubjectTarget())
+            .setStatus(ACTIVE)
+            .setContextTarget(encounter)
+            .setContext(new Reference(encounter))
+            .setPeriod(encounter.getPeriod());
+
+        getSupportingInfoFromITK(carePlan, cpSection);
+
+        return carePlan;
+    }
+
+    private void getSupportingInfoFromITK(CarePlan carePlan, POCDMT000002UK01Section carePlanSection) {
+        carePlansCreated.stream()
+                .filter(ofTypes(QuestionnaireResponse.class, Observation.class))
+                .map(Resource::getIdElement)
+                .map(Reference::new)
+                .forEach(carePlan::addSupportingInfo);
+    }
+
+    private Reference create(CarePlan carePlan) {
+        carePlansCreated.add(carePlan);
+        return new Reference(carePlan);
+    }
+
+    private List<POCDMT000002UK01Section> findCarePlanSections(POCDMT000002UK01Section section) {
+
+        // Base case: If there are no nested sub components
+        if (ArrayUtils.isEmpty(section.getComponentArray())) {
```

Comment:
maybe to use section.sizeOfComponentArray() instead?

Reply:
Applied the change accordingly